cuML/src/cuML.hpp

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include <cuda_runtime.h>
+
+namespace ML {
+
+class deviceAllocator {
```

Comment:
Does it make sense to reuse the `DeviceAllocator` defined inside ml-prims here: https://github.com/rapidsai/cuml/blob/branch-0.6/ml-prims/src/device_allocator.h? Or, even upgrading this existing class' interface?

Reply:
I very much agree to this. This should not be duplicated. The advantage of the interface proposed in this pull request is that it allows asynchronous allocation and deallocation (if supported by the allocator).

Reply:
I guess for now we can keep this duplication.
Once this PR gets through, we shall think about how to get rid of these duplications. (maybe we can just update the API of the ml-prims' `DeviceAllocator` itself?)

Reply:
I agree that we should do that after this is merged. I think the question how to avoid the duplication comes down to the question where the interface should live. Is my understanding correct that cuML has dependencies to ml-prims but not the other way around? In this case the Abstract base class definition of `ML::deviceAllocator` and `ML::hostAllocator` in `cuML.hpp` should simply be moved to a new hedaer in ml-prims that is then included from `cuML.hpp`.

Reply:
Yes, your understanding is correct.
If you can take care of this change in this PR itself, that'd be great!

Reply:
Done I moved `deviceAllocator` and `hostAllocator` to `cuml/ml-prims/src/common/cuml_allocator.hpp`. Please resolve this discussion if you are OK with this in the scope of this PR.